viewing_party/party.py

```diff
@@ -1,23 +1,320 @@
+    """
+    Add a movie to user's watchlist
+    
+    Parameters:
+        user_data (dict): with a key "watchlist" 
+        and values is a list of dictionaries movies user wants to watch
+        or empty list represent no movies in watchlist
+        movie (dict): {
+                "title": "Title A",
+                "genre": "Horror",
+                "rating": 3.5
+            }
+    
+    Return:
+        user_data (watchlist of movies)
+    """
+    user_data["watchlist"].append(movies)
+    return user_data
+ 
+    
+def watch_movie(user_data, title):
+    """
+    Parameters:
+        user_data (dict): with two keys "watchlist" and "watched" 
+        
+        title (str): represent title of movie user has watched
+    Return:
+        if title is in a movie in user's watchlist:
+            remove from watchlist
+            add to watched
+            return user_data
+        if not in watchlist:
+            return user_data
+    """  
+    for movie in user_data["watchlist"]:
+        if title == movie["title"]:
+            user_data["watchlist"].remove(movie) 
+            user_data["watched"].append(movie)
+        
+    return user_data
+
 
-# -----------------------------------------
 # ------------- WAVE 2 --------------------
 # -----------------------------------------
+def get_watched_avg_rating(user_data):
+    """
+    function calculate and return average rating
+    
+    Parameters:
+        user_data (dict): "watched" list of movie dictionaries.
+        
+    Notes:
+        if average rating of empty watched list if 0.0
+    
+    Returns:
+        Average rating (float)        
+    """
+    total_rating = 0
+    movies = 0
+    for movie in user_data["watched"]:
+        movies += 1
+        total_rating += movie["rating"]
+        
+    if movies > 0:    
+        average_rating = total_rating/movies   
+    else:
+        average_rating = 0.0   
+
+    return average_rating
+
+def get_most_watched_genre(user_data):
+    """
+    Get the most frequently watched genre from user's watched movies
+    
+    Parameter:
+        user_data (dict): 
+            the key "watched" is a list of dictionaries, where each dictionary represent a movie
+            each movie dict contains a key "genre" (str)
+    Returns:
+        str: the genre that user watched frequently
+        None: if "watched" list is empty
+    """
+    genres_count = {} # map from genre to count
+
+    # Return None if there is no value in "watched."
+    if user_data["watched"] == []:
+        return None
+    
+    # Add and count the genres found. 
+    for movie in user_data["watched"]:
+        genre = movie["genre"]
+        if genre not in genres_count:
+            genres_count[genre] = 1
+        else:
+            genres_count[genre] += 1
+            
+    max_rate = 0
+    most_frequent_genre = None
+    # Comparing the rates and returning the most frequenty genre.
+    for key, rate in genres_count.items():

```

Comment:
Small nitpick here, but if you're going to use `rate` for the value of each key, I would suggest using a more descriptive name than `key`!